AX: Implement NextParagraphEnd and PreviousParagraphStart off the main thread

hoffmanjoshua opened the pull request:
#### 2be2f2d783081daf22256d444a61df63f215f86c
<pre>
AX: Implement NextParagraphEnd and PreviousParagraphStart off the main thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=283112">https://bugs.webkit.org/show_bug.cgi?id=283112</a>
<a href="https://rdar.apple.com/139880052">rdar://139880052</a>

Reviewed by Tyler Wilcock.

This PR finishes moving paragraph navigation off the main thread by implementing the nextParagraphEnd
and previousParagraphStart APIs off of the main thread.

Several changes were made since 286004@main (AX: Implement TextUnit::Paragraph off the main thread),
including:
* Using the runIndex instead of the line index, so we can index relative to the current runs, not the
containing block.
* Adding logic to AXTextMarker::previousParagraphStart and AXTextMarker::nextParagraphEnd to mimic the
live tree behavior, notably, advancing when there are line breaks.
* Updating AXTextMarker::ParagraphRange to call findMarker directly, similar to the live tree.
* Refining the logic within findMarker to handle more edge cases. This includes removing the cumulative
offset, which is no longer used.

The text-marker-paragraph-nav.html test was also updated to test these APIs.

* LayoutTests/accessibility/ax-thread-text-apis/text-marker-paragraph-nav-expected.txt:
* LayoutTests/accessibility/ax-thread-text-apis/text-marker-paragraph-nav.html:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::previousParagraphStart const):
(WebCore::AXTextMarker::nextParagraphEnd const):
(WebCore::AXTextMarker::paragraphRange const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::previousParagraphStart const): Deleted.
(WebCore::AXTextMarker::nextParagraphEnd const): Deleted.
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper textMarkerForTextMarker:atUnit:]):

Canonical link: <a href="https://commits.webkit.org/286613@main">https://commits.webkit.org/286613@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14ac491a628d9c272fc01db6af96a06713f7118c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76468 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55503 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29374 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80994 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27744 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78585 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64645 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3796 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59958 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18071 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79535 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49871 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65665 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40286 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47266 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23154 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26067 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68392 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23486 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82441 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3844 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2527 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68233 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3997 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65634 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67478 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16839 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11455 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9544 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3791 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6600 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3814 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7244 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5572 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -937,24 +937,25 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
-        size_t startLineIndex = currentRuns->lineID(runIndex).lineIndex;
-        unsigned offsetInStartLine =  startLineIndex ? offset() - currentRuns->runLengthSumTo(startLineIndex) : offset();
+        unsigned sumToRunIndex = runIndex ? currentRuns->runLengthSumTo(runIndex - 1) : 0;
+        unsigned offsetInStartLine =  runIndex ? offset() - sumToRunIndex : offset();
```
`runIndex` seems to be `size_t`. Do we need to make this subtraction conditional? If it's 0, we'll just subtract offset() - 0, which is the same as just returning `offset()`.

hoffmanjoshua replied:
Ahh good point, this was from when I used to use startLineIndex instead. I'll update this one

twilco reviewed (approved):
Great job!

hoffmanjoshua committed:
AX: Implement NextParagraphEnd and PreviousParagraphStart off the main thread
https://bugs.webkit.org/show_bug.cgi?id=283112
rdar://139880052

Reviewed by Tyler Wilcock.

This PR finishes moving paragraph navigation off the main thread by implementing the nextParagraphEnd
and previousParagraphStart APIs off of the main thread.

Several changes were made since 286004@main (AX: Implement TextUnit::Paragraph off the main thread),
including:
* Using the runIndex instead of the line index, so we can index relative to the current runs, not the
containing block.
* Adding logic to AXTextMarker::previousParagraphStart and AXTextMarker::nextParagraphEnd to mimic the
live tree behavior, notably, advancing when there are line breaks.
* Updating AXTextMarker::ParagraphRange to call findMarker directly, similar to the live tree.
* Refining the logic within findMarker to handle more edge cases. This includes removing the cumulative
offset, which is no longer used.

The text-marker-paragraph-nav.html test was also updated to test these APIs.

* LayoutTests/accessibility/ax-thread-text-apis/text-marker-paragraph-nav-expected.txt:
* LayoutTests/accessibility/ax-thread-text-apis/text-marker-paragraph-nav.html:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::previousParagraphStart const):
(WebCore::AXTextMarker::nextParagraphEnd const):
(WebCore::AXTextMarker::paragraphRange const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::previousParagraphStart const): Deleted.
(WebCore::AXTextMarker::nextParagraphEnd const): Deleted.
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper textMarkerForTextMarker:atUnit:]):

Canonical link: https://commits.webkit.org/286613@main
